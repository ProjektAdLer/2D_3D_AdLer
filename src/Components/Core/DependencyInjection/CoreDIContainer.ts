import { LoadWorldController } from "../Presentation/React/LoadWorldButton/LoadWorldController";
import ILoadWorldController from "../Presentation/React/LoadWorldButton/ILoadWorldController";
import { Container } from "inversify";
import CORE_TYPES from "./CoreTypes";
import ISceneView from "../Presentation/Babylon/SceneManagement/ISceneView";
import SceneView from "../Presentation/Babylon/SceneManagement/SceneView";
import SceneViewModel from "../Presentation/Babylon/SceneManagement/SceneViewModel";
import ScenePresenter from "../Presentation/Babylon/SceneManagement/ScenePresenter";
import MainScene from "../Presentation/Babylon/SceneManagement/MainScene";
import ICreateSceneClass from "../Presentation/Babylon/SceneManagement/ICreateSceneClass";
import IScenePresenter from "../Presentation/Babylon/SceneManagement/IScenePresenter";
import infrastructureDIContainer from "./InfrastructureDIContainer";
import useCaseDIContainer from "./UseCases/UseCaseDIConatiner";
import PortsDIContainer from "./Ports/PortsDIContainer";
import BuilderDIContainer from "./Builders/BuilderDIContainer";
import INavigation from "../Presentation/Babylon/Navigation/INavigation";
import Navigation from "../Presentation/Babylon/Navigation/Navigation";
import NavigationConfiguration from "../Presentation/Babylon/Navigation/NavigationConfiguration";

var CoreDIContainer = new Container();

// Scene
CoreDIContainer.bind<ISceneView>(CORE_TYPES.ISceneView)
  .to(SceneView)
  .inSingletonScope();
CoreDIContainer.bind(SceneViewModel).toSelf().inSingletonScope();
CoreDIContainer.bind<IScenePresenter>(CORE_TYPES.IScenePresenter)
  .to(ScenePresenter)
  .inSingletonScope();

// Navigation
CoreDIContainer.bind<INavigation>(CORE_TYPES.INavigation)
  .to(Navigation)
  .inSingletonScope();
CoreDIContainer.bind(NavigationConfiguration).toSelf();

// Controllers
CoreDIContainer.bind<ILoadWorldController>(CORE_TYPES.ILoadWorldController)
  .to(LoadWorldController)
  .inSingletonScope();

// bind other CreateSceneClass here for testing puposes -MK
CoreDIContainer.bind<ICreateSceneClass>(CORE_TYPES.ICreateSceneClass).to(
  MainScene
);

CoreDIContainer.load(infrastructureDIContainer);
CoreDIContainer.load(useCaseDIContainer);
CoreDIContainer.load(PortsDIContainer);
CoreDIContainer.load(BuilderDIContainer);

export default CoreDIContainer;
