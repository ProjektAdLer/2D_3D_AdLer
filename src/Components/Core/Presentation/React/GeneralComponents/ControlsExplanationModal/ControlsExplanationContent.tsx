import { useTranslation } from "react-i18next";
import tailwindMerge from "../../../Utils/TailwindMerge";
import { AdLerUIComponent } from "src/Components/Core/Types/ReactTypes";

import leftClickIcon from "src/Assets/icons/left-click.svg";
import middleMouseIcon from "src/Assets/icons/middle-mouse-click.svg";
import middleMouseDragIcon from "src/Assets/icons/middle-mouse-click-and-drag.svg";
import touchIcon from "src/Assets/icons/touch.svg";
import touchDragIcon from "src/Assets/icons/touch-and-drag.svg";
import touchZoomIcon from "src/Assets/icons/touch-zoom.svg";
import hintIcon from "src/Assets/icons/hint.svg";

import LearningElementUnsolvedIcon from "src/Assets/icons/HighlightColors/LearningElementUnsolvedIcon.svg";
import LearningElementSolvedIcon from "src/Assets/icons/HighlightColors/LearningElementSolvedIcon.svg";
import NonLearningElementBaseIcon from "src/Assets/icons/HighlightColors/NonLearningElementBaseIcon.svg";

export default function ControlsExplanationContent({
  className,
}: AdLerUIComponent<{}>) {
  const { t: translate } = useTranslation("controls");
  return (
    <div
      className={tailwindMerge(
        className,
        "grid grid-rows-6 grid-flow-row gap-x-8 gap-y-3 pb-4 text-xs max-w-[85vw] mobile-landscape:justify-center mobile-landscape:grid-cols-1 mobile-landscape:grid-rows-auto mobile-landscape:grid-flow-row mobile-landscape:gap-y-3 mobile-landscape:pb-0 mobile-landscape:text-2xs mobile-landscape:gap-x-1 portrait:grid-flow-row portrait:grid-cols-1 portrait:grid-rows-auto portrait-gap-y-1 lg:text-sm xl:gap-2",
      )}
    >
      {/* Avatar Controls */}
      <h3 className="self-center font-bold text-adlerdarkblue">
        {translate("controlAvatar")}
      </h3>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Linksklick" src={leftClickIcon}></img>
        <p>{translate("controlAvatarMouse")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Einfacher Touch" src={touchIcon}></img>
        <p>{translate("controlAvatarTouch")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2 portrait:pb-3 ">
        <img className="w-6" alt="Icon" src={hintIcon}></img>
        <p>{translate("controlAvatarHint")}</p>
      </div>

      {/* Open Content Controls */}
      <h3 className="self-center font-bold text-adlerdarkblue">
        {translate("openContent")}
      </h3>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={leftClickIcon}></img>
        <p>{translate("openContentMouse")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={touchIcon}></img>
        <p>{translate("openContentTouch")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2 portrait:pb-3">
        <img className="w-6" alt="Icon" src={hintIcon}></img>
        <div className="flex flex-col gap-1">
          <p>{translate("openContentHint1")}</p>
          <p>{translate("openContentHint2")}</p>
        </div>
      </div>

      {/* Rotate Camera Controls */}
      <h3 className="self-center font-bold text-adlerdarkblue">
        {translate("rotateCamera")}
      </h3>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={middleMouseDragIcon}></img>
        <p>{translate("rotateCameraMouse")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={touchDragIcon}></img>
        <p>{translate("rotateCameraTouch")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2 portrait:pb-3">
        <img className="w-6" alt="Icon" src={hintIcon}></img>
        <p>{translate("rotateCameraHint")}</p>
      </div>

      {/* Zoom Camera Controls */}
      <h3 className="self-center font-bold text-adlerdarkblue">
        {translate("zoomCamera")}
      </h3>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={middleMouseIcon}></img>
        <p>{translate("zoomCameraMouse")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={touchZoomIcon}></img>
        <p>{translate("zoomCameraTouch")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2 portrait:pb-3">
        <img className="w-6" alt="Icon" src={hintIcon}></img>
        <p>{translate("zoomCameraHint")}</p>
      </div>

      {/* Highlight Colors */}
      <h3 className="self-center font-bold text-adlerdarkblue">
        {"Highlight Farben"}
      </h3>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={LearningElementUnsolvedIcon}></img>
        <p>{translate("LearningElementUnsolvedIcon")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={LearningElementSolvedIcon}></img>
        <p>{translate("LearningElementSolvedIcon")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={NonLearningElementBaseIcon}></img>
        <p>{translate("NonLearningElementBaseIcon")}</p>
      </div>
      <div className="flex flex-row items-start justify-start gap-2">
        <img className="w-6" alt="Icon" src={hintIcon}></img>
        <p>{translate("HighlightHint")}</p>
      </div>
    </div>
  );
}
