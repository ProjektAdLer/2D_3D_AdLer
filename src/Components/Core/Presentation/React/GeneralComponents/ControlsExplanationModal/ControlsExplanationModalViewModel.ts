import Observable from "src/Lib/Observable";

export default class ControlsExplanationModalViewModel {
  isOpen: Observable<boolean> = new Observable<boolean>(false);
}
