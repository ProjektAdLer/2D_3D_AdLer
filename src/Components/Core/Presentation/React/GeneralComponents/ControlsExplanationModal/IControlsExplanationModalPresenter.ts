export default interface IControlsExplanationModalPresenter {
  openModal(): void;
}
