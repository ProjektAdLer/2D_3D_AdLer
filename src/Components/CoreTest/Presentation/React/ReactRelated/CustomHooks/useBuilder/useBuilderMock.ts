import * as useBuilder from "~ReactComponents/ReactRelated/CustomHooks/useBuilder";

/**
 * Mocks the useViewModelControllerProvider hook.
 * @param mockedTupels The return value, which the mocked hook will return.
 *
 */
export default function useBuilderMock(
  mockedTupels: [viewModel: any, controller: any] = [undefined, undefined]
) {
  jest.spyOn(useBuilder, "default").mockReturnValue(mockedTupels);
}
