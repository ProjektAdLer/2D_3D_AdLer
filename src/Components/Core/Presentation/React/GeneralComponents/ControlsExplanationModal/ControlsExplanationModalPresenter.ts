import IControlsExplanationModalPresenter from "./IControlsExplanationModalPresenter";
import ControlsExplanationModalViewModel from "./ControlsExplanationModalViewModel";

export default class ControlsExplanationModalPresenter
  implements IControlsExplanationModalPresenter
{
  constructor(private viewModel: ControlsExplanationModalViewModel) {}

  openModal(): void {
    this.viewModel.isOpen.Value = true;
  }
}
