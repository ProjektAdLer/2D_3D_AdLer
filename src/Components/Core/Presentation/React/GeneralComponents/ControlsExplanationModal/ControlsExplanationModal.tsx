import ControlsExplanationModalViewModel from "./ControlsExplanationModalViewModel";
import useBuilder from "~ReactComponents/ReactRelated/CustomHooks/useBuilder";
import BUILDER_TYPES from "~DependencyInjection/Builders/BUILDER_TYPES";
import StyledModal from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledModal";
import { useTranslation } from "react-i18next";
import { AdLerUIComponent } from "src/Components/Core/Types/ReactTypes";
import tailwindMerge from "../../../Utils/TailwindMerge";
import useObservable from "~ReactComponents/ReactRelated/CustomHooks/useObservable";
import ControlsExplanationContent from "./ControlsExplanationContent";

export default function ControlsExplanationModal({
  className,
}: AdLerUIComponent<{}>) {
  const [viewModel] = useBuilder<ControlsExplanationModalViewModel, undefined>(
    BUILDER_TYPES.IControlsExplanationModalBuilder,
  );
  const { t: translate } = useTranslation("controls");

  const [isOpen, setIsOpen] = useObservable(viewModel?.isOpen);

  if (!viewModel) return null;
  if (!isOpen) return null;

  return (
    <StyledModal
      className={tailwindMerge(className, "")}
      title={translate("title")!}
      showModal={isOpen}
      onClose={() => {
        setIsOpen(false);
      }}
    >
      <ControlsExplanationContent
        className={tailwindMerge(
          className,
          "portrait:max-w-[90vw] max-h-[80vh] max-w-6xl",
        )}
      />
    </StyledModal>
  );
}
