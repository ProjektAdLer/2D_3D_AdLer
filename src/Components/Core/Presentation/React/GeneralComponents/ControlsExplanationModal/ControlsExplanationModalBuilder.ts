import { injectable } from "inversify";
import ControlsExplanationModalPresenter from "./ControlsExplanationModalPresenter";
import IControlsExplanationModalPresenter from "./IControlsExplanationModalPresenter";
import ControlsExplanationModalViewModel from "./ControlsExplanationModalViewModel";
import PresentationBuilder from "../../../PresentationBuilder/PresentationBuilder";
import CoreDIContainer from "~DependencyInjection/CoreDIContainer";
import PRESENTATION_TYPES from "~DependencyInjection/Presentation/PRESENTATION_TYPES";

@injectable()
export default class ControlsExplanationModalBuilder extends PresentationBuilder<
  ControlsExplanationModalViewModel,
  undefined,
  undefined,
  IControlsExplanationModalPresenter
> {
  constructor() {
    super(
      ControlsExplanationModalViewModel,
      undefined,
      undefined,
      ControlsExplanationModalPresenter
    );
  }

  buildPresenter(): void {
    super.buildPresenter();
    // ensure that previous instances of the presenter are unbound, when changing between spaces
    if (
      CoreDIContainer.isBound(
        PRESENTATION_TYPES.IControlsExplanationModalPresenter
      )
    )
      CoreDIContainer.unbind(
        PRESENTATION_TYPES.IControlsExplanationModalPresenter
      );

    CoreDIContainer.bind<IControlsExplanationModalPresenter>(
      PRESENTATION_TYPES.IControlsExplanationModalPresenter
    ).toConstantValue(this.presenter!);
  }
}
