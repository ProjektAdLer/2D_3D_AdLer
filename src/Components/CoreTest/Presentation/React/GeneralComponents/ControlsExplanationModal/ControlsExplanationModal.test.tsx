import { act, render } from "@testing-library/react";
import React from "react";
import ControlsExplanationModal from "../../../../../Core/Presentation/React/GeneralComponents/ControlsExplanationModal/ControlsExplanationModal";
import { Provider } from "inversify-react";
import CoreDIContainer from "../../../../../Core/DependencyInjection/CoreDIContainer";
import useBuilderMock from "../../ReactRelated/CustomHooks/useBuilder/useBuilderMock";
import ControlsExplanationModalViewModel from "../../../../../Core/Presentation/React/GeneralComponents/ControlsExplanationModal/ControlsExplanationModalViewModel";

describe("ControlsExplanationModal", () => {
  beforeEach(() => {
    const viewModel = new ControlsExplanationModalViewModel();
    viewModel.isOpen.Value = true;
    useBuilderMock([viewModel, undefined]);
  });

  test("doesn't render without view model", () => {
    useBuilderMock([undefined, undefined]);
    const { container } = render(
      <Provider container={CoreDIContainer}>
        <ControlsExplanationModal />
      </Provider>,
    );

    expect(container.firstChild).toBeNull();
  });

  // ANF-ID: [EZZ0033]
  test("should render", () => {
    const { container } = render(
      <Provider container={CoreDIContainer}>
        <ControlsExplanationModal />
      </Provider>,
    );

    expect(container).not.toBeEmptyDOMElement();
    expect(container).toMatchSnapshot();
  });

  test("doesn't render after the close button is clicked", () => {
    const { container, getByRole } = render(
      <Provider container={CoreDIContainer}>
        <ControlsExplanationModal />
      </Provider>,
    );
    const closeButton = getByRole("button", { hidden: true });

    act(() => {
      closeButton.click();
    });

    expect(container).toBeEmptyDOMElement();
  });
});
