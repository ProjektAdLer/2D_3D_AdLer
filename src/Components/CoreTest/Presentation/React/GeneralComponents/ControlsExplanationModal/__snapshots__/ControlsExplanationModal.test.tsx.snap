// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsExplanationModal should render 1`] = `
<div>
  <div
    class="z-50 flex items-center justify-center flex gap-2 m-2 w-fit h-fit text-adlertextgrey"
  >
    <div
      class="fixed top-0 bottom-0 left-0 right-0 flex items-center justify-center h-full bg-blacktrans"
    >
      <div
        class="z-50 p-2 rounded-lg bg-gradient-to-br from-adlerbggradientfrom to-adlerbggradientto max-w-[95%] max-h-[95%] overflow-hidden"
      >
        <div
          class="flex items-center justify-center w-full gap-2 p-1 pb-3 text-xl font-bold text-adlerdarkblue lg:roboto-black lg:text-2xl h-fit"
        >
          <div
            class="w-full"
          >
            3D-Steuerung
          </div>
          <div
            class=""
          >
            <button
              class="justify-center p-1 md:w-12 md:h-12 sm:w-10 sm:h-10 w-8 h-8 aspect-square flex items-center text-sm rounded-lg hover:cursor-pointer hover:border-buttonbgblue hover:text-buttonbgblue hover:bg-adlerdarkblue lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue"
            >
              <img
                alt="CloseButton"
                class="lg:w-10 md:w-8 sm:w-6"
                src="close-icon-nobg.svg"
              />
            </button>
          </div>
        </div>
        <div
          class="px-1 overflow-auto rounded-lg font-regular h-fit"
        >
          <div
            class="w-[90vw] grid grid-cols-4 portrait:grid-cols-1 gap-12 pb-4 text-xs lg:text-sm overflow-y-auto"
          >
            <div
              class="flex flex-col gap-4"
            >
              <h3
                class="font-bold"
              >
                Avatar bewegen
              </h3>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Linksklick auf eine Position
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Einfacher Touch auf eine Position
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Der Avatar kann nur innerhlab des mit Fussboden begrenzten Raums bewegt werden
                </p>
              </div>
            </div>
            <div
              class="flex flex-col gap-4"
            >
              <h3
                class="font-bold"
              >
                Inhalte öffnen
              </h3>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Linksklick auf ein 3D Objekt
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Einfacher Touch auf ein 3D Objekt
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Nur farblich umrandete 3D-Modelle mit 3D-Icons und Türen sind mit Inhalten verknüpft 
 Inhalte können nur geöffnet werden, wenn der Avatar sich in der Nähe des 3D-Objekts befindet   
                </p>
              </div>
            </div>
            <div
              class="flex flex-col gap-4"
            >
              <h3
                class="font-bold"
              >
                Kamera drehen
              </h3>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Mausrad drücken und seitliche Bewegung
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Einfacher Touch und seitliche Bewegung
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Die Kamera ist um 45° nach links und rechts schwenkbar
                </p>
              </div>
            </div>
            <div
              class="flex flex-col gap-4"
            >
              <h3
                class="font-bold"
              >
                Kamera zoomen
              </h3>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Mausrad drehen
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Zwei Touchpunkte entfernen / annäheren
                </p>
              </div>
              <div
                class="flex flex-row gap-2"
              >
                <img
                  alt="Icon"
                  class="w-6"
                />
                <p>
                  Die Kamera ist in einem bestimmen Nah- und Fernbereich zoombar
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
